nni/contrib/distillation/preprocessor.py

```diff
@@ -0,0 +1,184 @@
+        origin_collate_fn = dataloader.collate_fn
+
+        def collate_fn(data):
+            uids, origin_data = list(zip(*data))
+            return uids, origin_collate_fn(origin_data)  # type: ignore
+
+        setattr(dataloader, 'collate_fn', collate_fn)
+        return dataloader
+
+    def _patched_label_dataloader(self, labels_collate_fn: Callable[[List], Any]):
+        dataloader = self._patched_uid_dataloader()
+
+        uid_collate_fn = dataloader.collate_fn
+
+        def collate_fn(data):
+            uids, origin_batch = uid_collate_fn(data)
+            distil_labels = list(map(self._storage.select, cast(List[str], uids)))
+            distil_labels = labels_collate_fn(distil_labels)
+            return distil_labels, origin_batch
+
+        setattr(dataloader, 'collate_fn', collate_fn)
+        return dataloader
+
+    def generate_distillation_labels(self, epochs: int = 1, log_process: bool = True, labels_split_fn: Callable[[Any], List] | None = None):
+        """
+        Run the `teacher_predict` on the whole dataset, and record the output as distillation labels.
+
+        Parameters
+        ----------
+        epochs
+            By default, 1. It specifies how many epochs used to generate distillation labels, usually it can be set to 1.
+            And in some training methods, such as data augumentation, each time the same meta data selected from dataset
+            might be different after augumentation.
+            At this time, setting `>1` can run more epochs to obtain multiple epochs of distillation labels.
+        log_process
+            If logging during generate labels.
+        labels_split_fn
+            A function used to split the batched output of `teacher_predict` to list of output. If `labels_split_fn` is None,
+            `teacher_predict` is expected to return a batched tensor, and it will be auto split on dim 0 by this function.
+            For example, `teacher_predict` returns a tensor with size (32, 100), which is the probability of 100 classes for 32 samples,
+            then after `labels_split_fn`, it will be split to a list of tensor with 32 elements,
+            each tensor is the probability with size (100,).
+        """
+        if self._preprocessed:
+            warn_msg = 'Preprocessor has been generated distillation labels, if more epochs are needed, please set '
+            _logger.warning(warn_msg)
+            return
+
+        labels_split_fn = labels_split_fn if labels_split_fn is not None else self._labels_split_fn
+
+        self._dataset.observe()
+
+        dataloader = self._patched_uid_dataloader()
+        total_batch_num = len(dataloader)
+        log_literal = min(max(total_batch_num // 25, 10), 100)
+
+        for epoch_idx in range(epochs):
+            for batch_idx, packed_batch in enumerate(dataloader):
+                uids, batch = packed_batch
+                soft_labels = self._teacher_predict(batch)
```

Comment:
maybe teacher_model is better?

Reply:
in fact, here we need a predict function, not a model